File: src/gazap-panel/src/main/java/gazap/panel/web/mvc/oauth/OAuth1Flow.java
package gazap.panel.web.mvc.oauth;

import org.springframework.social.connect.Connection;
import org.springframework.social.connect.support.OAuth1ConnectionFactory;
import org.springframework.social.oauth1.*;

import javax.security.sasl.AuthenticationException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import java.io.IOException;

public class OAuth1Flow extends OAuthFlow<OAuth1ConnectionFactory<?>> {
    private static final String SESSION_REQUEST_TOKEN = "__oauth_flow1_request_token__";

    public OAuth1Flow(boolean initiationRequired, OAuth1ConnectionFactory<?> connectionFactory) {
        super(initiationRequired, connectionFactory);
    }

    @Override
    public void initiateAuthentication(HttpServletRequest request, HttpServletResponse response) throws IOException {
        OAuth1Operations operations = getConnectionFactory().getOAuthOperations();

        OAuthToken requestToken = operations.fetchRequestToken(getCallbackUrl(), null);
        storeCachedRequestToken(request, requestToken);

        OAuth1Parameters oauthParameters = OAuth1Parameters.NONE;
        if (operations.getVersion() == OAuth1Version.CORE_10) {
            oauthParameters = new OAuth1Parameters();
            oauthParameters.setCallbackUrl(getCallbackUrl());
        }
        response.sendRedirect(operations.buildAuthenticateUrl(requestToken.getValue(), oauthParameters));
    }

    @Override
    public Connection<?> attemptAuthentication(HttpServletRequest request) throws AuthenticationException {
        OAuthToken requestToken = extractCachedRequestToken(request);
        if (requestToken == null) {
            throw new AuthenticationException("lost request token");
        }

        String token = request.getParameter("oauth_token");
        String verifier = request.getParameter("oauth_verifier");

        OAuth1Operations operations = getConnectionFactory().getOAuthOperations();
        OAuthToken accessToken = operations.exchangeForAccessToken(new AuthorizedRequestToken(requestToken, verifier), null);
        return getConnectionFactory().createConnection(accessToken);
    }

    private void storeCachedRequestToken(HttpServletRequest request, OAuthToken token) {
        request.getSession().setAttribute(SESSION_REQUEST_TOKEN, token);
    }

    private OAuthToken extractCachedRequestToken(HttpServletRequest request) {
        HttpSession session = request.getSession(false);
        if (session == null) {
            return null;
        }
        try {
            return (OAuthToken) session.getAttribute(SESSION_REQUEST_TOKEN);
        } finally {
            session.removeAttribute(SESSION_REQUEST_TOKEN);
        }
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/web/mvc/oauth/OAuth1Flow.kt
package waypalm.site.web.mvc.oauth

import org.springframework.social.connect.support.OAuth1ConnectionFactory
import javax.servlet.http.HttpServletResponse
import javax.servlet.http.HttpServletRequest
import org.springframework.social.oauth1.OAuth1Parameters
import org.springframework.social.oauth1.OAuth1Version
import org.springframework.social.oauth1.AuthorizedRequestToken;
import org.springframework.social.connect.Connection
import javax.security.sasl.AuthenticationException
import org.springframework.social.oauth1.OAuthToken

public class OAuth1Flow(
        initiationRequired: Boolean,
        connectionFactory: OAuth1ConnectionFactory<*>,
        callbackUrl: String
)
: OAuthFlow<OAuth1ConnectionFactory<*>>(initiationRequired, connectionFactory, callbackUrl)
{
    class object {
        var SESSION_REQUEST_TOKEN: String = "__oauth_flow1_request_token__";
    }

    public override fun initiateAuthentication(request: HttpServletRequest, response: HttpServletResponse) {
        val operations = connectionFactory.getOAuthOperations()!!;

        val requestToken = operations.fetchRequestToken(callbackUrl, null)!!;
        storeCachedRequestToken(request, requestToken);

        var oauthParameters = OAuth1Parameters.NONE;
        if (operations.getVersion() == OAuth1Version.CORE_10) {
            oauthParameters = OAuth1Parameters();
            oauthParameters.setCallbackUrl(callbackUrl);
        }
        response.sendRedirect(operations.buildAuthenticateUrl(requestToken.getValue(), oauthParameters));
    }

    public override fun attemptAuthentication(request: HttpServletRequest): Connection<*>? {
        val requestToken = extractCachedRequestToken(request);
        if (requestToken == null) {
            throw AuthenticationException("lost request token");
        }

        val token = request.getParameter("oauth_token");
        val verifier = request.getParameter("oauth_verifier");

        val operations = connectionFactory.getOAuthOperations()!!;
        val accessToken = operations.exchangeForAccessToken(AuthorizedRequestToken(requestToken, verifier), null);
        return connectionFactory.createConnection(accessToken);
    }

    private fun storeCachedRequestToken(request: HttpServletRequest, token: OAuthToken) {
        request.getSession()!!.setAttribute(SESSION_REQUEST_TOKEN, token);
    }

    private fun extractCachedRequestToken(request: HttpServletRequest): OAuthToken? {
        val session = request.getSession(false);
        if (session == null) {
            return null;
        }
        try {
            return session.getAttribute(SESSION_REQUEST_TOKEN) as OAuthToken;
        } finally {
            session.removeAttribute(SESSION_REQUEST_TOKEN);
        }
    }

}